server/util/boundedstack/boundedstack.go

```diff
@@ -0,0 +1,100 @@
+package boundedstack
+
+import (
+	"sync"
+
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+)
+
+// BoundedStack is a concurrency-safe stack with bounded size: when the capacity
+// of the stack is exceeded, the least recently added item ("bottom") of the
+// stack is removed.
+//
+// Its intended usage is to prioritize the most recently added items for eager
+// chunk fetching.
+//
+// It exposes a channel C which streams a value whenever a value is pushed on
+// the stack.
+//
+// IMPORTANT: after receiving a message from C, make sure to always call Pop.
+// Otherwise, the stack item corresponding to the notification on C may be
+// dropped unnecessarily.
+//
+// Example code for a goroutine that processes the latest items pushed to the
+// stack:
+//
+//	for range stack.C {
+//		val, ok := stack.Pop()
+//		if !ok {
+//			// A receive from C normally guarantees that a value is ready,
+//			// but it's best to always check `ok` when popping from the stack.
+//			continue
+//		}
+//		// Do something with val...
+//	}
+type BoundedStack[T any] struct {
+	// C is a notification channel which allows goroutines to listen for values
+	// being pushed to the stack.
+	//
+	// After a goroutine receives from C, it must call Pop - otherwise the stack
+	// item corresponding to the notification on C may be dropped unnecessarily.
+	C chan struct{}
+
+	mu     sync.Mutex
+	buffer []T
+	top    int
+	size   int
+}
+
+func New[T any](capacity int) (*BoundedStack[T], error) {
+	if capacity <= 0 {
+		return nil, status.InvalidArgumentError("bounded stack capacity must be > 0")
+	}
+	return &BoundedStack[T]{
+		C:      make(chan struct{}, capacity),
+		buffer: make([]T, capacity),
+	}, nil
+}
+
+// Push adds an item to the top of the stack, evicting the bottom item of the
+// stack if the stack is full.
+func (s *BoundedStack[T]) Push(item T) {
+	s.push(item)
+	// Notify the channel that a new item was pushed.
+	select {
+	case s.C <- struct{}{}:
+	default:
+		// If the channel buffer is full then it just means that the goroutines
+		// processing values from this stack aren't keeping up; this is fine.
+	}
+}
+
+func (s *BoundedStack[T]) push(item T) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.top = (s.top + 1) % len(s.buffer)
+	s.buffer[s.top] = item
+	s.size = min(s.size+1, len(s.buffer))
+}
+
+// Pop returns the top element of the stack. The second return value indicates
+// whether the returned item is valid; it will be false if and only if the stack
+// was empty.
+func (c *BoundedStack[T]) Pop() (item T, ok bool) {
```

Comment:
Should we make the channel a private implementation detail of pop? That way clients don't have to worry about `IMPORTANT: after receiving a message from C, make sure to always call Pop.
// Otherwise, the stack item corresponding to the notification on C may be
// dropped unnecessarily.`

Pop could be a blocking method

Reply:
+1 the API is kind of error prone as written

Reply:
I considered this but didn't want to make `Pop()` a blocking method because otherwise goroutines could get stuck indefinitely without some other cancellation / closing mechanism which also felt error prone.

I agree that the channel approach is not ideal though so I added a `Recv(ctx)` method to allow cancellation if the context is canceled, and made the channel private. wdyt? (Went with the name "recv" since it's analogous to a receive on a channel, and because we might want to reserve the name "pop" for the non-blocking pop() operation in case it makes sense to have that be public later).

Reply:
Should we make `Pop` private? It seems incorrect that `Push` will always add an element to the channel, but `Pop` won't remove one. You could just add a large comment that `Recv` is a blocking method?

Reply:
Made it private. I was using it for tests to check that Pop() returns nothing when the stack is empty, and it would have been awkward to do that via Recv instead.